### ISSUE TITLE
Just used in production

### ISSUE THREAD

**Issue by atakanyenel** [julvo/tinystatic#1]
Hello, I saw the app on Reddit. I just used it to migrate a website from php to pure html. I was using the php just for including partial html like header and footer so this project fit perfectly to the task. It was very easy to use and gives great control to user. Here are some remarks that might be useful:

- I think there should be a `partial` command for embedding partial into routes. Currently they can only be embedded in templates. My index page is unique, no need for templating, but it uses partials, so it would be very nice to have that command. I ended up defining a index_temp.html as a template with all the content and created my index.html in routes just with:

```
---
template: index_temp.html
---
```
For that workaround I had to create 2 files with similar names, so there is that.

- My website had 2 templates, and 6 routes in total. It was very fast for me to render them with this. One improvement would be to put your generate loop in main.go in a goroutine. Those tasks can run in parallel, I guess, so your code would be something like:
```go
for _, r := range allRoutes {
       go func(route Route){
                if route.Href != "" {
			log.Println("∟", route.FilePath, "->", route.Href)
		}
		if err := route.Generate(outputDir, allRoutes); err != nil {
			log.Fatalln(err)
		}
         }(r)
	}
```
you can also get rid of `outputDir` & `allRoutes` parameters for the generate method. They are already known and constant so your AllRoutes() method can take them and return routes array with those information already embedded in them.

---

I really liked the simplicity of the project. I didn't need a autoload server at all. Keeping a python server running in output and refreshing the page was enough. It really does not worth the effort of creating websockets for hot reloading.

A future milestone would be to minimize generated output. Basically removing whitespace from html and minifying assets. Hugo does that, but hugo is utterly complicated. So this is your call.

**Comment by julvo** [comment]
Hey @atakanyenel, thanks for giving tinystatic a try and taking the time to share this feedback - I really appreciate it. 

Good point about the partials. Using partials without using a page template could be very helpful. I'll add that soon, without changing the API. Instead of simply copying the HTML if no template is specified, `{{template "..." .}}` pipelines should be evaluated. In the meantime, you could also create a template `empty.html` with just `{{template "body" .}}` and then use that template in your `index.html` by specifying it in the meta data and wrapping the content in `{{define "body"}} ... {{end}}`. That way you could keep the content in `routes` instead of `templates`, but your solution works equally well.

I like the idea to parallelise the generation, even if it's probably fast enough for me for now. Will have a look at the difference between serial and parallel and see what there is to gain.

Also good idea to minify assets. It's something I've been looking at as well and will add that soon.

Thanks again for the feedback!

**Comment by julvo** [comment]
@atakanyenel Just made the change to use the template engine, even when no template is specified, so that you can use partials in routes without templates.

I also tested parallel generation. The generation time for the small blog example went from ~125ms to ~110ms on average. Before including that into master, I'd like to test this more in real world use and see if performance ever is an issue. If not, I'd rather not introduce parallel generation, as that could lead to confusing log/error interleaving without making more changes.

**Comment by atakanyenel** [comment]
Hey, that's great, I can help generating a big mock project for better testing. Partial part is great too.

An issue I realised also was that , when `template:default.html`, without any space between `:` and the template name, the project does not render. It's very easy to fix on the user part but the error message is not informative.

**Comment by julvo** [comment]
If you like, feel free to open a PR for adding a bigger example project to the examples folder in the repo.

Good point regarding the error messages, they do need some more work to become user friendly. In this specific case you mentioned, the problem is probably that if a file begins with `---`, tinystatic expects there to be a yaml object at the top of the file. Without the space after the colon, it's not a yaml object, but simply a yaml string of even invalid yaml. In this case parsing the yaml fails and currently, there's probably some generic yaml parsing error, rather than a more user friendly alternative. Thanks for bringing this up!

**Comment by julvo** [comment]
Closing this for now, feel free to comment or create more issues if anything comes up. Thanks again for your feedback and ideas!
